primitives/src/v1.rs

```diff
@@ -476,3 +479,221 @@ pub struct AvailableData {
 	/// The omitted validation data.
 	pub omitted_validation: OmittedValidationData,
 }
+
+/// A helper data-type for tracking validator-group rotations.
+#[derive(Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(PartialEq, Debug))]
+pub struct GroupRotationInfo<N = BlockNumber> {
+	/// The block number where the session started.
+	pub session_start_block: N,
+	/// How often groups rotate. 0 means never.
+	pub group_rotation_frequency: N,
+	/// The current block number.
+	pub now: N,
+}
+
+impl GroupRotationInfo {
+	/// Returns the index of the group needed to validate the core at the given index, assuming
+	/// the given number of cores.
+	///
+	/// `core_index` should be less than `cores`, which in turn should be less than u32::max().
```

Comment:
```suggestion
	/// `core_index` should be less than `cores`, which in turn should be less than or equal to `u32::max()`.
```
or
```suggestion
	/// `core_index` should be less than `cores`, which is capped at `u32::max()`.
```

Reply:
actually 4 billion cores sounds like too many, I wonder if it should be limited by millions instead

Reply:
this is an arithmetic limit, not a practical one. I expect that we in practice won't have more than a few thousand cores.

Reply:
`Vec::<u8>::with_capacity(usize::max())` will fail even though, while legal under API, it is not a practical amount of memory to allocate.